numpy/core/src/umath/ufunc_object.c

```diff
@@ -1748,10 +1763,153 @@ make_arr_prep_args(npy_intp nin, PyObject *args, PyObject *kwds)
     }
 }
 
+/*
+ * Check whether any of the outputs of a gufunc has core dimensions.
+ */
+static int
+_has_output_coredims(PyUFuncObject *ufunc) {
+    int i;
+    for (i = ufunc->nin; i < ufunc->nin + ufunc->nout; ++i) {
+        if (ufunc->core_num_dims[i] > 0) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/*
+ * Interpret a possible axes keyword argument, using it to remap the default
+ * axes (at the end of the operands) to new ones.
+ *
+ * Returns 0 on success, and -1 on failure
+ */
+static int
+_parse_axes_arg(PyUFuncObject *ufunc, PyObject *axes, PyArrayObject **op,
+                int broadcast_ndim, int **remap_axis) {
+    int nin = ufunc->nin;
+    int nout = ufunc->nout;
+    int nop = nin + nout;
+    int i, list_size;
+
+    if (!PyList_Check(axes)) {
+        PyErr_SetString(PyExc_TypeError, "axes should be a list.");
+        return -1;
+    }
+    list_size = PyList_Size(axes);
+    if (list_size != nop) {
+        if (list_size != nin || _has_output_coredims(ufunc)) {
+            PyErr_Format(PyExc_ValueError,
+                         "axes should be a list with an entry for all "
+                         "%d inputs and outputs; entries for outputs can only "
+                         "be omitted if none of them has core axes.",
+                         nop);
+            return -1;
+        }
+        for (i = nin; i < nop; i++) {
+            remap_axis[i] = NULL;
+        }
+    }
+    for (i = 0; i < list_size; ++i) {
+        int j, iter_axis;
+        int op_ndim, core_ndim, num_axes, item_is_tuple;
+        int requested_axis = 0;
+        int have_seen_axis[NPY_MAXDIMS] = {0};
+        PyObject *item;
+
+        core_ndim = ufunc->core_num_dims[i];
+        if (op[i] != NULL) {
+            op_ndim = PyArray_NDIM(op[i]);
+        }
+        else {
+            op_ndim = broadcast_ndim + core_ndim;
+        }
+        item = PyList_GetItem(axes, i);
+        item_is_tuple = PyTuple_Check(item);
+        if (item_is_tuple) {
```

Comment:
I'd be tempted here to just do
```c
if (!item_is_tuple) {
    item = PyTuple_Pack(1, item);
}
else {
    Py_INCREF(item);
}
```
and then you don't need to special case the shorthand further down - this shouldn't be the bottleneck.

Reply:
I was dithering about this, but agree the code simplicity wins here, so I do the tuple creation now.